classes/ThankYous/Api.php

```diff
@@ -561,14 +574,124 @@ public function GetThankYouUrlById(int $id)
+	 *
+	 * @param array $data
 	 * @return ThankYou
+	 * @throws ValidationException - If the Thank You could not be created from the parameter provided.
+	 * @throws ThankYouNotFoundException - If the Thank You could not be found.
 	 */
-	public function Create($author, string $description, ?Date $date_created = null)
+	public function CreateFromArray(array $data)
 	{
-		return $this->thank_yous_repository->Create($author, $description, $date_created);
+		$id          = $data['id'] ?? null;
+		$author      = $data['author'] ?? null;
+		$description = $data['description'] ?? null;
+
+		$thanked = $data['thanked'] ?? null;
+		$tag_ids = $data['tags'] ?? null;
+
+		$errors = [];
+
+		if (isset($id))
+		{
+			$thank_you = $this->GetThankYou($id);
+			if (isset($description))
+			{
+				$thank_you->SetDescription($description);
+			}
+		} else
+		{
+			if (!isset($description))
+			{
+				$errors[] = ['name' => 'description', 'reason' => ($this->lmsg)('thankyou.thankyou.description.error.empty')];
+			}
+
+			if (!isset($author))
+			{
+				$errors[] = ['name' => 'author', 'reason' => ($this->lmsg)('thankyou.thankyou.author.error.undefined')];
+			} elseif (!is_int($author) && !($author instanceof User))
+			{
+				$errors[] = ['name' => 'author', 'reason' => ($this->lmsg)('thankyou.thankyou.author.error.invalid')];
+			}
+
+			if (!empty($errors))
+			{
+				throw new ValidationException($errors);
+			}
+
+			$thank_you = $this->thank_you_factory->Create($author, $description, null);
+
+			if (isset($id) && !is_int($id))
+			{
+				$errors[] = ['name' => 'id', 'reason' => ($this->lmsg)('thankyou.thankyou.id.error.invalid')];
+			} else
+			{
+				$thank_you->SetId($id);
+			}
+		}
+
+		if (isset($thanked))
+		{
+			if (!is_array($thanked))
+			{
+				$errors[] = ['name' => 'thanked', 'reason' => ($this->lmsg)('thankyou.thanked.error.invalid')];
+			} else
+			{
+				foreach ($thanked as $offset => $oclass)
+				{
+					$thanked[$offset] = ['oclass' => (int) ($oclass['oclass'] ?? null), 'id' => (int) ($oclass['id'] ?? null)];
+				}
+				try
+				{
+					$thankables = $this->CreateThankablesFromOClasses($thanked);
+					$thank_you->SetThanked($thankables);
+				} catch (UnsupportedOwnerClassException $exception)
+				{
+					$errors[] = [
+						'name'   => 'thanked',
+						'reason' => ($this->lmsg)('thankyou.thankable.owner_class.error.not_supported',
+							implode(', ', $this->GetThankableObjectTypes())
+						)
+					];
+				}
+			}
+		}
+
+		if (isset($tag_ids))
+		{
+			foreach ($tag_ids as $offset => $tag_id)
+			{
+				if (!is_int($tag_id))
+				{
+					$invalid_params[] = [
+						'name'   => 'tags',
+						'reason' => ($this->lmsg)('thankyou.tag.error.id.invalid', (string) $tag_id)
+					];
+					unset($tag_ids[$offset]);
+				}
+			}
+			$tags = $this->tag_api->GetTagsById($tag_ids);
+			foreach ($tag_ids as $tag_id)
+			{
+				if (!isset($tags[$tag_id]))
+				{
+					$invalid_params[] = [
```

Comment:
Looks like this should be `$errors` instead of `$invalid_params`, same as the loop above. This variable is otherwise unused.

Reply:
Fixed in 4dc232252644d7d5a766bb9d648c6831bab4e4e0